### PR TITLE
Implement vpx (VP8 and VP9) encoder

### PR DESCRIPTION
**Pull request by at-wat**
Requires libvpx.

### CODE REVIEW

File: pkg/codec/vpx/vpx.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		}
+		return n, err
+	}
+
+	img, err := e.r.Read()
+	yuvImg := img.(*image.YCbCr)
+
+	e.raw.planes[0] = (*C.uchar)(&yuvImg.Y[0])
+	e.raw.planes[1] = (*C.uchar)(&yuvImg.Cb[1])
+	e.raw.planes[2] = (*C.uchar)(&yuvImg.Cr[2])
```
Comment (lherman-cs) [inline pkg/codec/vpx/vpx.go]: I think they should be `&yuvImg.Cb[0]` and `&yuvImg.Cr[0]`

**Comment by lherman-cs** [comment]
Btw, the color seems to be off. But, I really don't mind it. This is definitely a good start! And, I'm sorry for the mess...

Screenshot from 2020-02-08 10-20-10

File: pkg/codec/vpx/vpx.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+			e.raw.fmt = C.VPX_IMG_FMT_I444
+		case image.YCbCrSubsampleRatio422:
+			e.raw.fmt = C.VPX_IMG_FMT_I422
+		case image.YCbCrSubsampleRatio420:
+			e.raw.fmt = C.VPX_IMG_FMT_I420
+		case image.YCbCrSubsampleRatio440:
+			e.raw.fmt = C.VPX_IMG_FMT_I440
+		default:
+			return 0, fmt.Errorf("unsupported pixel format: %s", yuvImg.SubsampleRatio)
+		}
```
Comment (lherman-cs) [inline pkg/codec/vpx/vpx.go]: Is this probably the reason why the color is off? Just curious.

Comment (at-wat) [reply pkg/codec/vpx/vpx.go]: VP8 looks only supporting I420 and YV12. (VP9 supports I422, I444 and I440 on profile1, but pion/rtp doesn't have payloader yet.)
On my environment, image from camera is I422.

I'll add convert to make it work.

File: pkg/codec/vpx/vpx.go (hunk 1 above)
Comment (at-wat) [reply pkg/codec/vpx/vpx.go]: oops, fixed.

**Comment by at-wat** [comment]
@lherman-cs fixed the pointer and the color problem. Please take another look.

File: pkg/codec/vpx/vpx.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	e.raw.planes[0] = (*C.uchar)(&yuvImg.Y[0])
+	e.raw.planes[1] = (*C.uchar)(&yuvImg.Cb[0])
+	e.raw.planes[2] = (*C.uchar)(&yuvImg.Cr[0])
+	e.raw.stride[0] = C.int(yuvImg.YStride)
+	e.raw.stride[1] = C.int(yuvImg.CStride)
+	e.raw.stride[2] = C.int(yuvImg.CStride)
+	h := yuvImg.Rect.Max.Y - yuvImg.Rect.Min.Y
+
+	// Covert pixel format to I420
+	// TODO: maybe better to move it to a proper module
```
Comment (lherman-cs) [inline pkg/codec/vpx/vpx.go]: Definitely agree. It looks like we can put this under `io/video` in the future. The pattern will look like a middleware. For example,

```
func ToI420(r video.Reader) video.Reader
```

File: pkg/codec/vpx/vpx.go (hunk 2 above)
Comment (lherman-cs) [reply pkg/codec/vpx/vpx.go]: Ah, I see.

File: pkg/codec/vpx/vpx.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		for i := 0; i < h/2; i++ {
+			addrSrc := i * 2 * yuvImg.CStride
+			addrDst := i * yuvImg.CStride
+			for j := 0; j < yuvImg.CStride; j++ {
+				cb := uint16(yuvImg.Cb[addrSrc+j]) + uint16(yuvImg.Cb[addrSrc+yuvImg.CStride+j])
+				cr := uint16(yuvImg.Cr[addrSrc+j]) + uint16(yuvImg.Cr[addrSrc+yuvImg.CStride+j])
+				yuvImg.Cb[addrDst+j] = uint8(cb / 2)
+				yuvImg.Cr[addrDst+j] = uint8(cr / 2)
+			}
+		}
```
Comment (lherman-cs) [inline pkg/codec/vpx/vpx.go]: Wow, I didn't expect that would implement them now. Thanks a lot ^^

**Pull request merged**